nix/keys.nix

```diff
@@ -123,32 +147,41 @@ in
       "things to break."
     )];
 
+    assertions = flip mapAttrsToList config.deployment.keys (key: opts: {
+      assertion = (opts.text == null && opts.keyFile != null) ||
+                  (opts.text != null && opts.keyFile == null);
+      message = "Deployment key '${key}' must have either a 'text' or a 'keyFile' specified.";
+    });
+
     system.activationScripts.nixops-keys = stringAfter [ "users" "groups" ]
       ''
         mkdir -p /run/keys -m 0750
         chown root:keys /run/keys
 
         ${optionalString config.deployment.storeKeysOnMachine
-            (concatStrings (mapAttrsToList (name: value:
-              let
-                # FIXME: The key file should be marked as private once
-                # https://github.com/NixOS/nix/issues/8 is fixed.
-                keyFile = pkgs.writeText name value.text;
-              in "ln -sfn ${keyFile} /run/keys/${name}\n")
-              config.deployment.keys)
+            (concatStrings (mapAttrsToList
+                            (name: value: let
+                                            # FIXME: The key file should be marked as private once
+                                            # https://github.com/NixOS/nix/issues/8 is fixed.
+                                            keyFile = if !isNull value.keyFile
+                                                      then value.keyFile
+                                                      else pkgs.writeText name value.text;
+                                          in "ln -sfn ${keyFile} /run/keys/${name}\n")
+                           config.deployment.keys)
             + ''
               # FIXME: delete obsolete keys?
               touch /run/keys/done
             '')
         }
 
-        ${concatStringsSep "\n" (flip mapAttrsToList config.deployment.keys (name: value:
-          # Make sure each key has correct ownership, since the configured owning
-          # user or group may not have existed when first uploaded.
-          ''
-            [[ -f "/run/keys/${name}" ]] && chown '${value.user}:${value.group}' "/run/keys/${name}"
-          ''
-        ))}
+        ${optionalString (!config.deployment.storeKeysOnMachine)
```

Comment:
@domenkozar, because calling `chown` on a symlink pointing to read-only Nix store failed, when I was testing it -- I was actually surprised how it worked before.